std-bits/table/src/main/java/org/enso/table/data/index/MultiValueKeyBase.java

```diff
@@ -0,0 +1,92 @@
+package org.enso.table.data.index;
+
+import org.enso.table.data.column.storage.Storage;
+
+/** The base class for keys used for sorting/grouping rows by a set of columns. */
+public abstract class MultiValueKeyBase {
+  protected final Storage[] storages;
+  protected final int rowIndex;
+  protected boolean hasFloatValues = false;
+  protected boolean floatsComputed = false;
+
+  /**
+   * Constructs a key based on an array of column storages and the index of the row the key is
+   * associated with.
+   */
+  public MultiValueKeyBase(Storage[] storage, int rowIndex) {
+    this.storages = storage;
+    this.rowIndex = rowIndex;
+  }
+
+  /** A helper function to get the item from the nth column of the key's row. */
+  protected Object get(int column) {
+    return storages[column].getItemBoxed(rowIndex);
+  }
+
+  @Override
+  public abstract boolean equals(Object o);
+
+  /** Checks if all cells in the current row are missing. */
+  public boolean areAllNull() {
+    for (Storage value : storages) {
+      if (!value.isNa(rowIndex)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /* Checks if any cell contains float values.
+
+   It takes value folding into account, i.e. a float value that can be coerced to an integer without loss of precision is not considered floating.
+  */
+  public boolean hasFloatValues() {
+    if (!floatsComputed) {
```

Comment:
Up to you, that was also my initial intuition and I did what I did just to comply with the tests that were there. It is debatable since in Enso `2 == 2.0` so they are practically the same. But my personal opinion would be to warn for `2.0` anyway - the algorithm is giving us doubles not ints (they are `==` but they are _still_ distinguishable, for example with matching) and this suggests something fishy may be going on - if you don't want warnings better be explicit and cast to int along the way.

So I'm all for changing the behaviour to warn on _any_ floats (the code will also be simpler :slightly_smiling_face:  ), so please just confirm and I can modify accordingly.

Reply:
Think any float makes sense.